service/storage_proxy.cc

```diff
@@ -1165,7 +1165,9 @@ class cas_mutation : public mutation_holder {
     };
 };
 
-class abstract_write_response_handler : public seastar::enable_shared_from_this<abstract_write_response_handler> {
+using is_cancellable = bool_class<struct cancellable_tag>;
+
```

Comment:
This raises the question: should we cancel all outstanding writes, not just hints? Nothing guarantees that regular writes don't have long timeouts too. But that's outside the scope of the series.

Reply:
One reason not to do that is user experience. If we're shutting down gracefully, then for example any regular QUORUM read that has been already routed to this replica should finish normally. If we cancelled it then the coordinator will have to do a speculative retry which would increase latency or worse, the read will fail.

Reply:
Yes. We should drain requests that are being waited on. But should we wait for requests that were acknowledged but are still active? Probably it doesn't matter that we do, since if the remote node is down, it will be detected and the request will complete.

Why doesn't this happen for hints?

Reply:
> Why doesn't this happen for hints?

You mean detecting that a node is down and cancelling the hint write? It does after this PR
```cpp
void storage_proxy::on_down(const gms::inet_address& endpoint) {
    return cancel_write_handlers([endpoint] (const abstract_write_response_handler& handler) {
        const auto& targets = handler.get_targets();
        return boost::find(targets, endpoint) != targets.end();
    });
};

```

Reply:
I was confused. The patch generalizes the cancel functionality from view handlers to all write handlers, and enables it for hint handlers. Too many similar things.